### ISSUE TITLE
PHP 8.1

### ISSUE THREAD

**Issue by mmenavas** [frictionlessdata/tableschema-php#70]
# Overview

Given that [PHP 8.0 EOL](https://www.php.net/supported-versions.php) is on November 2023, I'd like to know if there's any ongoing effort to make this library compatible with php 8.1.

Below are the results of running [PHPStan](https://phpstan.org/) agains the `src` directory. Since I'm working on a project that uses this library, I'd be more than happy to help out with this effort.

```
 ------ ----------------------------------------------------------------------------------- 
  Line   Fields/BaseField.php                                                               
 ------ ----------------------------------------------------------------------------------- 
  :101   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
  :118   Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
 ------ ----------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------- 
  Line   Schema.php                                                                         
 ------ ----------------------------------------------------------------------------------- 
  :76    Unsafe usage of new static().                                                      
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static  
 ------ ----------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------- 
  Line   SchemaValidator.php                                                                         
 ------ -------------------------------------------------------------------------------------------- 
  :28    Access to an undefined property frictionlessdata\tableschema\SchemaValidator::$descriptor.  
  :29    Access to an undefined property frictionlessdata\tableschema\SchemaValidator::$errors.      
  :37    Access to an undefined property frictionlessdata\tableschema\SchemaValidator::$errors.      
  :52    Access to an undefined property frictionlessdata\tableschema\SchemaValidator::$errors.      
  :59    Access to an undefined property frictionlessdata\tableschema\SchemaValidator::$descriptor.  
  :78    Access to an undefined property frictionlessdata\tableschema\SchemaValidator::$descriptor.  
 ------ -------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   Table.php                                                                                                                                                   
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  :57    Unsafe usage of new static().                                                                                                                               
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static                                                                           
  :163   Return type mixed of method frictionlessdata\tableschema\Table::current() is not covariant with tentative return type mixed of method Iterator::current().  
         💡 Make it covariant, or use the #[\ReturnTypeWillChange] attribute to temporarily suppress the error.                                                      
  :216   Return type mixed of method frictionlessdata\tableschema\Table::rewind() is not covariant with tentative return type void of method Iterator::rewind().     
         💡 Make it covariant, or use the #[\ReturnTypeWillChange] attribute to temporarily suppress the error.                                                      
  :226   Return type mixed of method frictionlessdata\tableschema\Table::key() is not covariant with tentative return type mixed of method Iterator::key().          
         💡 Make it covariant, or use the #[\ReturnTypeWillChange] attribute to temporarily suppress the error.                                                      
  :231   Return type mixed of method frictionlessdata\tableschema\Table::next() is not covariant with tentative return type void of method Iterator::next().         
         💡 Make it covariant, or use the #[\ReturnTypeWillChange] attribute to temporarily suppress the error.                                                      
  :238   Return type mixed of method frictionlessdata\tableschema\Table::valid() is not covariant with tentative return type bool of method Iterator::valid().       
         💡 Make it covariant, or use the #[\ReturnTypeWillChange] attribute to temporarily suppress the error.                                                      
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------ 
```

EDIT: Adding deprecated warnings issued by PHP 8.1:
```
PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 163

PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::next() should either be compatible with Iterator::next(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 231

PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::key() should either be compatible with Iterator::key(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 226

PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::valid() should either be compatible with Iterator::valid(): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 238

PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::rewind() should either be compatible with Iterator::rewind(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 216

```

---

Please preserve this line to notify @courtney-miles (lead of this repository)

**Comment by OriHoch** [comment]
Hi, sounds good, I think a PR would be very welcome, let me know if you need any help

**Comment by DiegoPino** [comment]
Hi Maximo, happy to work on more PHP version compatibility, but this would need to be a new release since some in-the-wild uses are still probably on older PHPs. We do use the Library on PHP 8.1 without errors so not sure if PHPStan is complaining or just giving warnings there. Will check

Will review my next week schedule (bit crazy these days) and see if I find some time. Pull requests are also welcome. Will update here

Best


Diego Pino Navarro
Director of Digital Strategy
Archipelago architect
Metropolitan New York Library Council
PO Box 2084
New York, NY 10108

**Comment by courtney-miles** [comment]
> We do use the Library on PHP 8.1 without errors

Admittedly, the known compatibility with PHP gets a little hairy because its version constraint is `>=7.1.0`.

We run the tests for all versions up to 8.0, but after that compatibility is an unknown.

As a part of ensuring compatibility with PHP 8.1, would we be happy to peg the upper version of PHP in `composer.json` so it can't be used with versions before we have verified compatibility?

**Comment by mmenavas** [comment]
@DiegoPino: I agree with you that the library works on PHP 8.1 without errors; however, I'm seeing `PHP Deprecated` messages like the ones below. I know I can suppress those messages, but I think this is an opportunity to contribute code to future proof the library.

```
PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 163
```
```
PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::next() should either be compatible with Iterator::next(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 231
```
```
PHP Deprecated:  Return type of frictionlessdata\tableschema\Table::key() should either be compatible with Iterator::key(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /Users/.../vendor/frictionlessdata/tableschema/src/Table.php on line 226
```

**Comment by mmenavas** [comment]
@courtney-miles: I created PR #71 to address the deprecated messages.

As a side note, it looks like we can ignore most of the PHPStan messages as PHP8.1 (and 8.2) are only yelling at us when the return types of overriding methods are omitted or incompatible with those of the parent methods.

**Comment by courtney-miles** [comment]
I'm sorry for coming back to this sooo late.

I have raised #72 because there was a number of changes to address deprecation notices.

The hairiest part with PHP 8.1 is that `strptime` is deprecated -- which means PHP has no native way to interpret the format pattern for [Date](https://specs.frictionlessdata.io/table-schema/#date) (and datetime, time.)  So those patterns have had to be transformed to the format that `date_parse_from_format` understands -- this is done with best effort, but then my understanding is that `strptime` was dropped because it itself cannot consistently interpret patterns between system.

**Comment by courtney-miles** [comment]
Changes have been released in version 1.2.0.
